### ISSUE TITLE
GetOrAdd extension method likely incorrectly appplying to ConcurrentDictionary

### ISSUE THREAD

**Issue by jaredpar** [xunit/xunit#2931]
The `DictionaryExtensions` type defines the following `GetOrAdd` method

```csharp
	public static TValue GetOrAdd<TKey, TValue>(
		this IDictionary<TKey, TValue> dictionary,
		TKey key,
		Func<TValue> newValue)
			where TKey : notnull
```

This means that it applies to `ConcurrentDictionary<TKey, TValue>` however the implementation of this extension method does not line up with the semantics of the actual `GetOrAdd` method. That method ensures that a key will only ever store a single value. The extension method though will override existing values if called concurrently. 

That extension method is being called instead of the actual instance method in [TestIntrospectionHelper](https://github.com/xunit/xunit/blob/main/src/xunit.v3.core/Sdk/Frameworks/TestIntrospectionHelper.cs#L25). That is because the instance method overload that takes a delegate requires the signature `Func<TKey, TValue>` where as the calls in TestIntrospectionHelper pass no arguments hence go to this extension method. 

Not sure this is causing an actual bug but seems incorrect.

**Comment by bradwilson** [comment]
This is in an internal class, not a public one, so how is it being called by your code? Is it coming from xunit.assert.source?

**Comment by bradwilson** [comment]
Just realized this isn't a problem in v2 because I had the names as "AddOrGet" instead of "GetOrAdd" to prevent such collisions. No idea why I "fixed" them in v3, as that seems like the wrong behavior.

**Comment by bradwilson** [comment]
And actually it _is_ a problem in v2 as well because there are places I was using AddOrGet that should've been using GetOrAdd. 🤦🏼‍♂️

**Comment by bradwilson** [comment]
Fixed in v2 `2.8.1-pre.14`
Fixed in v3 `0.1.1-pre.424`

Thanks!

**Comment by jaredpar** [comment]
> This is in an internal class, not a public one, so how is it being called by your code? 

Know you crossed this out but I happened on this code by accident. This afternoon I was alerted to a bug in [nuget][nuget] around `GetOrAdd`. Essentially believing this code pattern is calling the factory version of `GetOrAdd` but really it's calling the value overload.

```csharp
ConcurrentDictionary<string, object> table = new();
void M(string key) {
  // This calls GetOrAdd(string, object) not GetOrAdd(string, Func<string, object> valueFactory) because the 
  // lambda is missing an argument.
  var lockObj = table.GetOrAdd(key, () => new object()); 
  lock (lockObj) { 
    ...
```
It's a very subtle bug. I was concerned at how wide spread it could be so I was using [grep.app][grep] to search for that pattern. One of the hits was xunit. It didn't hit that bug (value parameter isn't `object`) but I was still confused at how the code was compiling. Eventually I found the extension method and noticed that the logic probably wasn't right for `ConcurrentDictionary`. 

Given I had gone that far figured I should go ahead and file a bug to let you know :smile: 

My searching did turn up another real instance of the bug in [msbuild][msbuild]. 

[nuget]: https://github.com/NuGet/NuGet.Client/pull/5808
[msbuild]: https://github.com/dotnet/msbuild/pull/10152
[grep]: https://grep.app/search?q=GetOrAdd%5C%28.%2A%2C%5Cs%2A%5C%28%5Cs%2A%5C%29&regexp=true&filter[lang][0]=C%23
